### PR TITLE
feat: add child account

### LINKED ISSUE #34: Account Hierarchy: As a B2B Admin User, I should be able to add child account
https://www.figma.com/file/bKJuIwUx6VXmubHZo4rCBq/B2B?type=design&node-id=457-25616&mode=design&t=n8BQuXy14qz46hIH-0

```
mutation  {
  createCustomerB2bAccount (b2BAccountInput: {
      id: 100001,
      parentAccountId: 1004,
      companyOrOrganization: "Child 1",
      taxId: "123456",
      users: [
        {
          emailAddress: "sushant2005@gmail.com",
      		userName: "sushant2005@gmail.com",
     			firstName: "Jadhav",
          localeCode: "en-IN"
        }
      ]    
  }){
  	id
    parentAccountId
    companyOrOrganization
  }
}
```

### PR DESCRIPTION
**Pull request by bsatignitiv**

### CODE REVIEW

File: components/dialogs/AccountHierarchyFormDialog/AccountHierarchyFormDialog.tsx
Comment (kibo-chandradeeptalaha) [inline components/dialogs/AccountHierarchyFormDialog/AccountHierarchyFormDialog.tsx:1]: Add this folder inside a b2b folder.

File: components/my-account/User/UserTable/UserTable.tsx
Comment (kibo-chandradeeptalaha) [inline components/my-account/User/UserTable/UserTable.tsx:1]: can we add this inside of components/b2b ?

File: components/my-account/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx
Comment (kibo-chandradeeptalaha) [inline components/my-account/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx:1]: Can we add this inside Components/b2b ?

File: components/dialogs/AccountHierarchyFormDialog/AccountHierarchyFormDialog.tsx
Comment (bsatignitiv) [reply components/dialogs/AccountHierarchyFormDialog/AccountHierarchyFormDialog.tsx:1]: Done

File: components/my-account/User/UserTable/UserTable.tsx
Comment (bsatignitiv) [reply components/my-account/User/UserTable/UserTable.tsx:1]: Done

File: components/my-account/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx
Comment (bsatignitiv) [reply components/my-account/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx:1]: Done

File: __mocks__/stories/accountHierarchy.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  customerSet: 'default',
+  commerceSummary,
+  segments: [],
+  taxExempt: false,
+  taxId: '123456',
+  accountType: 'B2B',
+  migrationRequired: false,
+}
+
+export const accountHierarchy = {
```
Comment (kibo-sushant) [inline __mocks__/stories/accountHierarchy.ts]: user the Type
rename file with [TypeName]Mock

File: __mocks__/stories/createCustomerB2bAccountMock.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    ],
+  },
+}
+
+export const b2BAccountResponseMock: B2BAccount = {
+  id: 0,
+  parentAccountId: 1023,
+  taxId: '123234',
+  companyOrOrganization: 'Ignitiv Corp',
+}
```
Comment (kibo-sushant) [inline __mocks__/stories/createCustomerB2bAccountMock.ts]: create two separate files
name the files [TypeNanme]Mock

File: components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+  const onSubmit = () => {
+    if (isLoading || !isValid) return
+    setLoading(true)
+    const formValues = getValues()
+    onSave({
+      ...formValues,
+      parentAccount: selectedParentAccount,
+    })
+  }
```
Comment (kibo-sushant) [inline components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx:81]: After onSave where we are setting setLoading(false)?

File: components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      name: 'parent-account',
+    })
+
+    await waitFor(() =>
+      expect(companyNameField.value).toBe(
+        AddAccountToChild?.args?.accounts?.[0]?.companyOrOrganization
+      )
+    )
+  })
+})
```
Comment (kibo-sushant) [inline components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.spec.tsx:100]: missing stories
should call onSave callback function when user clicks on 'Create Account' button
should call onClose callback function when user clicks on 'Cancel' button

File: components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.spec.tsx [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    // expect(collapsedItems).toHaveLength(2) // Both items should be collapsed
+
+    // Click the Expand All button and verify that all items are expanded
+    // fireEvent.click(expandAllButton)
+    // const expandedItems = screen.getAllByTestId('expand-more-icon')
+    // expect(expandedItems).toHaveLength(2) // Both items should be expanded
+  })
+
+  // You can write more test cases to cover other scenarios as well
+})
```
Comment (kibo-sushant) [inline components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.spec.tsx:66]: missing testcases
should not show any action button to 'Non-Purchaser.'
should only show 'Buyer' and 'Quotes' link to 'Purchaser.'
should show all the links and action buttons to 'Admin'
should call onBuyersClick callback function when user ...
should call  onQuotesClick  callback function when user ...
should call   onAdd  callback function when user ...
should call  onEdit  callback function when user ...
should call   onDelete  callback function when user ...

File: components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.tsx [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+        handler={<Handler />}
+        renderCollapseIcon={({ isCollapsed }) => (
+          <CollapseStateIndicator isCollapsed={isCollapsed} />
+        )}
+        onChange={confirmChange}
+        //onChange={(items) => console.log(items)}
+      />
+    </RoleContext.Provider>
+  )
+}
```
Comment (kibo-sushant) [inline components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.tsx:91]: Don't put multiple components in single page, let's separate it out.
Write the Testcases for each one.
Mock all the children while testing

File: components/dialogs/b2b/AccountHierarchyFormDialog/AccountHierarchyFormDialog.spec.tsx [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+  it('should render save and cancel text on button when isUserFormInDialog is true', () => {
+    render(<Common {...Common.args} />)
+
+    const cancelText = screen.getByText('cancel')
+    expect(cancelText).toBeVisible()
+
+    const saveText = screen.getByText('create-account')
+    expect(saveText).toBeVisible()
+  })
+})
```
Comment (kibo-sushant) [inline components/dialogs/b2b/AccountHierarchyFormDialog/AccountHierarchyFormDialog.spec.tsx:89]: 'should render form title passed in props' it should be part of the  'should render component' testcases, don't create separate testcase

Missing Testcases
should call onSave callback function when user ...
should call onClose callback function when user...

File: components/page-templates/B2B/AccountHierarchyTemplate/AccountHierarchyTemplate.spec.tsx [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    })
+
+    setup()
+
+    const accountTitleElement = screen.getByText(/my-account/i)
+    fireEvent.click(accountTitleElement)
+
+    expect(mockPush).toHaveBeenCalledWith('/my-account')
+  })
+})
```
Comment (kibo-sushant) [inline components/page-templates/B2B/AccountHierarchyTemplate/AccountHierarchyTemplate.spec.tsx:153]: mock the AccountHierarchyFormDialog

File: hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+ * @returns 'response?.data?.createCustomerB2bAccount' which contains object of account added
+ */
+
+export const useCreateCustomerB2bAccountMutation = () => {
+  return {
+    createCustomerB2bAccount: useMutation({
+      mutationFn: createCustomerB2bAccount,
+    }),
+  }
+}
```
Comment (kibo-sushant) [inline hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts:43]: Don't we have to invalidate anything?

File: hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+import { useMutation } from '@tanstack/react-query'
+
+import { makeGraphQLClient } from '@/lib/gql/client'
+
+import { MutationCreateCustomerB2bAccountArgs } from '@/lib/gql/types'
+import { createCustomerB2bAccountMutation } from '@/lib/gql/mutations'
+
+const client = makeGraphQLClient()
+
+const createCustomerB2bAccount = async (b2BAccountInput: MutationCreateCustomerB2bAccountArgs) => {
```
Comment (kibo-sushant) [inline hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts]: apply return type

**Review by kibo-chandradeeptalaha — APPROVED** [review]

File: hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts (hunk 10 above)
Comment (bsatignitiv) [reply hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts]: Done

File: hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts (hunk 9 above)
Comment (bsatignitiv) [reply hooks/mutations/b2b/useCreateCustomerB2bAccount/useCreateCustomerB2bAccount.ts:43]: Yes, we have and its done in another PR

File: __mocks__/stories/accountHierarchy.ts (hunk 1 above)
Comment (bsatignitiv) [reply __mocks__/stories/accountHierarchy.ts]: Done

File: __mocks__/stories/createCustomerB2bAccountMock.ts (hunk 2 above)
Comment (bsatignitiv) [reply __mocks__/stories/createCustomerB2bAccountMock.ts]: Done

File: components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx (hunk 3 above)
Comment (bsatignitiv) [reply components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.tsx:81]: onSave method doesn't return the promise. 
setLoading(false) stops the loader immediately. 
By not using setLoading(false) loader is visible until the form is submitted. 
After successful response dialog closes so not required to set isLoading false.

File: components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.spec.tsx (hunk 4 above)
Comment (bsatignitiv) [reply components/b2b/AccountHierarchy/AccountHierarchyForm/AccountHierarchyForm.spec.tsx:100]: Added missing test cases

File: components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.tsx (hunk 6 above)
Comment (bsatignitiv) [reply components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.tsx:91]: Separated components with story component and test cases

File: components/dialogs/b2b/AccountHierarchyFormDialog/AccountHierarchyFormDialog.spec.tsx (hunk 7 above)
Comment (bsatignitiv) [reply components/dialogs/b2b/AccountHierarchyFormDialog/AccountHierarchyFormDialog.spec.tsx:89]: added missing test cases

File: components/page-templates/B2B/AccountHierarchyTemplate/AccountHierarchyTemplate.spec.tsx (hunk 8 above)
Comment (bsatignitiv) [reply components/page-templates/B2B/AccountHierarchyTemplate/AccountHierarchyTemplate.spec.tsx:153]: its already mocked with the name AccountHierarchyDialog
renamed it to AccountHierarchyFormDialog

File: components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.spec.tsx (hunk 5 above)
Comment (bsatignitiv) [reply components/b2b/AccountHierarchy/AccountHierarchyTree/AccountHierarchyTree.spec.tsx:66]: added missing test cases. covered above test cases as screen for Admin, Purchaser and Non-Purchaser

**Review by kibo-sushant — APPROVED** [review]

**Pull request merged**